### PR TITLE
Split literal constraints detection from `CanonicalizeMfp`.

### PR DESCRIPTION
**Pull request by ggevay**
We are planning to later call `CanonicalizeMfp` many more times, so we shouldn't do the expensive literal constraints detection along with canonicalizing an MFP. This PR splits off the literal constraint detection part of `CanonicalizeMfp` into a `LiteralConstraints` transform, and does a minor reordering in the optimizer pipeline as follows:
- `LiteralConstraints` is called instead of `CanonicalizeMfp` where `CanonicalizeMfp` was being called for the first time.
- `CanonicalizeMfp` is called for the first time in `logical_cleanup_pass`. 
  - This means that other transforms in `logical_cleanup_pass` can benefit from `CanonicalizeMfp`, see https://materializeinc.slack.com/archives/C02PPB50ZHS/p1673518074245349?thread_ts=1673463352.216739&cid=C02PPB50ZHS
  - (Later we will probably change it to be called after every transform as part of the normalization Frank proposed in the below-linked Slack thread.)
- The old second call of `CanonicalizeMfp` is just the remaining part of CanonicalizeMfp.

Note that it's beneficial for `LiteralConstraints` to have already had a `CanonicalizeMfp` recently. (Currently, `CanonicalizeMfp` is in `logical_cleanup_pass`, and then `LiteralConstraints` is right after that in `physical_optimizer`.)

### Motivation

   * This PR refactors existing code. See https://materializeinc.slack.com/archives/C02PPB50ZHS/p1673530709119299?thread_ts=1673463352.216739&cid=C02PPB50ZHS

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - Existing tests cover this.
  - The only slt change is in `not-null-propagation.slt`, and it looks like an improvement!
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Let's say none (just some minor incidental plan improvements in some exotic cases)

### CODE REVIEW

File: src/transform/src/literal_constraints.rs [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+            exprs: mfp.predicates.iter().map(|(_, p)| p.clone()).collect(),
+        }];
+        *mfp = MapFilterProject::new(mfp.input_arity)
+            .map(map)
+            .filter(new_predicates)
+            .project(project);
+    }
+
+    /// Call [mz_expr::canonicalize::canonicalize_predicates] on each of the predicates in the MFP.
+    fn canonicalize_predicates(mfp: &mut MapFilterProject, relation: &MirRelationExpr) {
```
Comment (frankmcsherry) [inline src/transform/src/literal_constraints.rs:499]: I'm reading through this, and the main confusion I have is what the bar is for inclusion in `literal_constraints.rs`. I did not read deeply, but superficially this name makes it sound like it it should be in `canonicalize_mfps.rs`? There are several "expression mechanics" in this file, and might (in the future, now, somewhere in between) they find a more specific home?

**Review by frankmcsherry — APPROVED** [review]
I like the principle, and I do like the outcomes also. It feels like there is still room to figure out how to chop up the supporting logic, maybe, as we have quite a few things that dance over expressions that don't seem like they are exclusively in support of translating literal equality into indexed look up. Or, maybe I'm wrong and it's just surprising.

File: src/transform/src/literal_constraints.rs (hunk 1 above)
Comment (ggevay) [reply src/transform/src/literal_constraints.rs:499]: The main call to `canonicalize_predicates` is in `fusion/filter.rs`. Now that you mention it, this indeed seems a bit strange. It's also strange that `CanonicalizeMfp` also does the actual fusion part of filter fusion. I created an issue to sort this out: "Merge filter fusion and `CanonicalizeMfp`" https://github.com/MaterializeInc/materialize/issues/17142

Comment (ggevay) [reply src/transform/src/literal_constraints.rs:499]: And the rest of the expression manipulation in `literal_constraints.rs` is strictly to support the literal constraints detection. (Btw. I have an issue to refactor literal constraints detection into a recursive implementation from the current disjunctive normal form-based implementation: https://github.com/MaterializeInc/materialize/issues/14797)

**Pull request merged**